src/Microsoft.AspNetCore.Mvc.ViewFeatures/Rendering/MvcForm.cs

```diff
@@ -63,27 +70,24 @@ protected virtual void GenerateEndForm()
         private void RenderEndOfFormContent()
         {
             var formContext = _viewContext.FormContext;
-            if (formContext.HasEndOfFormContent)
+            if (!formContext.HasEndOfFormContent)
             {
-                var writer = _viewContext.Writer;
-                var htmlWriter = writer as HtmlTextWriter;
+                return;
+            }
 
-                HtmlEncoder htmlEncoder = null;
-                if (htmlWriter == null)
+            var viewBufferWriter = _viewContext.Writer as ViewBufferTextWriter;
+            if (viewBufferWriter == null)
+            {
+                foreach (var content in formContext.EndOfFormContent)
                 {
-                    htmlEncoder = _viewContext.HttpContext.RequestServices.GetRequiredService<HtmlEncoder>();
+                    content.WriteTo(_viewContext.Writer, _htmlEncoder);
                 }
-
+            }
+            else
+            {
                 foreach (var content in formContext.EndOfFormContent)
                 {
-                    if (htmlWriter == null)
-                    {
-                        content.WriteTo(writer, htmlEncoder);
-                    }
-                    else
-                    {
-                        htmlWriter.Write(content);
-                    }
+                    viewBufferWriter.Write(content);
```

Comment:
I don't love having this special case here. We could consider bringing back `HtmlTextWriter` for this purpose, or just saying Razor + ViewBufferTextWriter and joined at the hip and shrug.